lib/src/conflict_labels.rs

```diff
@@ -0,0 +1,168 @@
+/// them more efficient to clone.
+#[derive(ContentHash, PartialEq, Eq, Clone)]
+pub struct ConflictLabels {
+    labels: Option<Arc<Merge<String>>>,
+}
+
+impl ConflictLabels {
+    /// Create a `ConflictLabels` with no labels.
+    pub const fn unlabeled() -> Self {
+        Self { labels: None }
+    }
+
+    /// Create a `ConflictLabels` from a `Merge<String>`. If the merge is
+    /// resolved or if any label is empty, the labels will be discarded, since
+    /// resolved merges cannot have labels, and labels cannot be empty.
+    pub fn new(labels: Merge<String>) -> Self {
+        if labels.is_resolved() || labels.iter().any(|label| label.is_empty()) {
+            Self::unlabeled()
+        } else {
+            Self {
+                labels: Some(Arc::new(labels)),
+            }
+        }
+    }
+
+    /// Create a `ConflictLabels` from a `Vec<String>`, with an empty vec
+    /// representing no labels.
+    pub fn from_vec(labels: Vec<String>) -> Self {
+        if labels.is_empty() {
+            Self::unlabeled()
+        } else {
+            Self::new(Merge::from_vec(labels))
+        }
+    }
+
+    /// Returns true if there are labels present.
+    pub fn is_present(&self) -> bool {
+        self.labels.is_some()
+    }
+
+    /// Returns the number of labeled sides, or `None` if unlabeled.
+    pub fn num_sides(&self) -> Option<usize> {
+        self.labels.as_ref().map(|labels| labels.num_sides())
+    }
+
+    /// Returns the underlying labels as an `Option<&Merge<String>>`.
+    pub fn as_merge(&self) -> Option<&Merge<String>> {
+        self.labels.as_ref().map(Arc::as_ref)
+    }
+
+    /// Returns the underlying labels as an `Option<Merge<String>>`, cloning if
+    /// necessary.
+    pub fn into_merge(self) -> Option<Merge<String>> {
+        self.labels.map(Arc::unwrap_or_clone)
+    }
+
+    /// Returns the conflict labels as a slice. If there are no labels, returns
+    /// an empty slice.
+    pub fn as_slice(&self) -> &[String] {
+        self.as_merge().map_or(&[], |labels| labels.as_slice())
+    }
+
+    /// Get the label for a side at an index.
+    pub fn get_add(&self, add_index: usize) -> Option<&str> {
+        self.as_merge()
+            .and_then(|merge| merge.get_add(add_index).map(String::as_str))
+    }
+
+    /// Get the label for a base at an index.
+    pub fn get_remove(&self, remove_index: usize) -> Option<&str> {
+        self.as_merge()
+            .and_then(|merge| merge.get_remove(remove_index).map(String::as_str))
+    }
+}
+
+impl From<Merge<String>> for ConflictLabels {
+    fn from(value: Merge<String>) -> Self {
+        Self::new(value)
+    }
+}
+
+impl From<Merge<&'_ str>> for ConflictLabels {
+    fn from(value: Merge<&str>) -> Self {
+        Self::new(value.map(|&label| label.to_owned()))
+    }
+}
+
+impl<T> From<Option<T>> for ConflictLabels
+where
+    T: Into<Self>,
+{
+    fn from(value: Option<T>) -> Self {
+        value.map_or_else(Self::unlabeled, T::into)
+    }
+}
```

Comment:
I often find named constructors clearer than `.into()`. Maybe that's just me, so feel free to ignore.